### ISSUE TITLE
I/Q swapped on bcrm43455c0 (RPi3/4)?

### ISSUE THREAD

**Issue by mzakharo** [seemoo-lab/nexmon_csi#44]
Hello,

The comment  in https://github.com/seemoo-lab/nexmon_csi/blob/b52fca3abc18715d6d12692e531164b5d62a78fd/src/csi_extractor.c#L219

Implies that the code below is placing imaginary into the lower 16 bit. 
 https://github.com/seemoo-lab/nexmon_csi/blob/b52fca3abc18715d6d12692e531164b5d62a78fd/src/csi_extractor.c#L224-L225 

complex numbers  are normally packed  with real in the lower address.

**Comment by jlinktu** [comment]
Hi,
extracted CSI, 32bit per subcarrier, hold the quadrature (imaginary) part in the lower 14bit and the in-phase (real) part in the 14bit above. We convert them to 16bit each for easier processing, keeping them in order. This I-Q-order is quite common, but if you prefer to switch real and imaginary part in your project you can easily do that without breaking anything.

**Comment by mzakharo** [comment]
Then isnt the sample matlab code incorrect then?
  Also,  switching downstream is doable, but not 'efficient'. Below code is vectorized conversion of 16 bit array to numpy:

`H = np.frombuffer(buffer, dtype='<h').astype(np.float32).view(np.complex64)`

Data being in Q/I order prevents usage of this vectorized API.

**Comment by jlinktu** [comment]
True. The order is wrong in the example for bcm4339 and bcm43455c0, and it makes more sense to have I/Q order in the UPD payload. For now you can just swap lines https://github.com/seemoo-lab/nexmon_csi/blob/b52fca3abc18715d6d12692e531164b5d62a78fd/src/csi_extractor.c#L222 and https://github.com/seemoo-lab/nexmon_csi/blob/b52fca3abc18715d6d12692e531164b5d62a78fd/src/csi_extractor.c#L224.
Will probably push it here soon too.

**Comment by Junye97** [comment]
Hi @jlinktu, does that imply in order to get the correct phase information, we will need to swap the real and imaginary component in MATLAB for processing?
